### PR TITLE
Introduce `with_forced_trimmed_paths`

### PR DESCRIPTION
**Pull request by estebank**
Built on top of https://github.com/rust-lang/rust/pull/104922, only last commit is relevant.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 1]
```diff
@@ -295,11 +297,85 @@ pub trait PrettyPrinter<'tcx>:
...
+                    .unwrap_or(name)
+            } else {
+                name
+            }
+        };
+        if let DefKind::Variant = kind
+            && let Some(symbol) = self.tcx().trimmed_def_paths(()).get(&def_id)
+        {
+            // If `Assoc` is unique, we don't want to talk about `Trait::Assoc`.
+            // self.write_str(symbol.as_str())?;
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/print/pretty.rs:330]: commented out code

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 2]
```diff
@@ -295,11 +297,85 @@ pub trait PrettyPrinter<'tcx>:
         self.try_print_visible_def_path_recur(def_id, &mut callers)
     }
 
+    fn force_print_trimmed_def_path(
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/print/pretty.rs:300]: document this function

**Review by oli-obk — COMMENTED** [review]
r=me after the PR this is based on has landed

**Comment by estebank** [comment]
Fixed single test divergence

@bors r=oli-obk

**Pull request merged**